allennlp_models/seq2seq/bart/bart.py

```diff
@@ -0,0 +1,401 @@
+        """
+        super().__init__(vocab)
+        self.bart = BartForConditionalGeneration.from_pretrained(model_name)
+
+        self._indexer = indexer or PretrainedTransformerIndexer(model_name, namespace="tokens")
+
+        self._start_id = self.bart.config.bos_token_id  # CLS
+        self._decoder_start_id = self.bart.config.decoder_start_token_id or self._start_id
+        self._end_id = self.bart.config.eos_token_id  # SEP
+        self._pad_id = self.bart.config.pad_token_id  # PAD
+
+        self._max_decoding_steps = max_decoding_steps
+        self._beam_search = BeamSearch(
+            self._end_id, max_steps=max_decoding_steps, beam_size=beam_size or 1
+        )
+
+        self._rogue = ROUGE(exclude_indices={self._start_id, self._pad_id, self._end_id})
+        self._bleu = BLEU(exclude_indices={self._start_id, self._pad_id, self._end_id})
+
+        # Replace bart encoder with given encoder. We need to extract the two embedding layers so that
+        # we can use them in the encoder wrapper
+        if encoder is not None:
+            assert (
+                encoder.get_input_dim() == encoder.get_output_dim() == self.bart.config.hidden_size
+            )
+            self.bart.model.encoder = _BartEncoderWrapper(
+                encoder,
+                self.bart.model.encoder.embed_tokens,
+                self.bart.model.encoder.embed_positions,
+            )
+
+    @overrides
+    def forward(
+        self, source_tokens: TextFieldTensors, target_tokens: TextFieldTensors = None
+    ) -> Dict[str, torch.Tensor]:
+        """
+        Performs the forward step of Bart.
+
+        # Parameters
+
+        source_tokens : `TextFieldTensors`, required
+            The source tokens for the encoder. We assume they are stored under the `tokens` key.
+        target_tokens : `TextFieldTensors`, optional (default = None)
+            The target tokens for the decoder. We assume they are stored under the `tokens` key. If no target
+            tokens are given, the source tokens are shifted to the right by 1.
+
+
+        # Returns
+
+        Dict[str, torch.Tensor]
+            During training, this dictionary contains the `decoder_logits` of shape `(batch_size,
+            max_target_length, target_vocab_size)` and the `loss`. During inference, it contains `predictions`
+            of shape `(batch_size, max_decoding_steps)` and `log_probabilities` of shape `(batch_size,)`.
+
+        """
+        inputs = source_tokens
+        targets = target_tokens
+        input_ids, input_mask = inputs["tokens"]["token_ids"], inputs["tokens"]["mask"]
+
+        outputs = {}
+
+        # TODO: (not targets) => generating?
+        # assert targets is None or self.training  # Don't provide targets for generation
+        generating = not self.training
+
+        # If no targets are provided, then shift input to right by 1. Bart already does this interally
+        # but it does not
```

Comment:
But it does not what?

Reply:
It doesn't use those for the loss calculation, I fixed the comment.